taperecorder QoL

NX115 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<strong>Single Tweak</strong>
I only edited one file: 'taperecorder.dm' & added a new sprite for the function. I added quality of life functionality to both the tape recorder and individual tapes for them to be wiped. I also added some quality-of-life functions so that both items are easier to interact with. This comes in the form to alt-clicking tapes to flip, and alt-clicking recorder to start/stop recording. 

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->


## Why It's Good For The Game

<strong>Increased RP Flexibility Through Convenience</strong>

I have recently attempted to play a 'Journalist' character and found that prolonged interviews are hampered by how many tapes you need to record the entire interview. This is a hassle for both finding enough tapes, carrying them, and managing them. 

The tape recorder has a transcript-printing functionality, which means once a tape is full a player can print out everything recorded, meaning there is no further reason to retain a tape that can no longer be used for recording. This allows for an easy way of recording and documenting events without fearing for the loss of content because you need to either make, find or swap tapes every 10 minutes.


<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: NX115 | NOXA
qol: tape & tape recorder smoother to utilize
add: the ability to wipe tapes 
sprite: added 'erase' for the tape recorder's 'wipe' function
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

NX115 committed:
taperecorder QoL

added the ability to wipe tapes through tape interaction and through the recorder.

Refactored the code so that it's using fewer lines by adding || instead of multiple ifs with the same outcome.

Changed the AltClick function of the tape recorder to make using it more efficient, instead of just playing it back.

ArcaneMusic commented:
I have a minor request and if it's outside of the scope of your PR then feel free to say no:

Could you make it possible to make pre-recorded tape messages? I've wanted to use tape messages left on ruins as a better storytelling tool over leaving sheets of paper all over the place for awhile, but never got around to it. It would especially be nice for more puzzles/lore on maps.

Fikou commented:
the refactor in your changelog is not relevant to players, remove it please

SuperNovaa41 commented:
I think we're supposed to avoid adding new verbs

Mothblocks reviewed (changes requested):
Revert your if-changes. I actively discourage the type of thing you're doing because it makes it harder to expand and harder to understand what is going on.

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -430,6 +433,26 @@
+		to_chat(usr, "<span class='notice'>You flip the tape back around.</span>")
+	tapeflip()
+
+/obj/item/tape/verb/flipVerb()
```
No more verbs, if you're up for it you can use a contextual screentip instead--you will want to check out https://github.com/tgstation/tgstation/tree/master/code/datums/elements/screentips#datumelementcontextual_screentip_bare_hands.

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -345,6 +334,7 @@
 	var/unspooled = FALSE
 	var/list/icons_available = list()
 	var/radial_icon_file = 'icons/hud/radial_tape.dmi'
+	var/firstFlip = TRUE
```
`first_flip`--snake_case only

NX115 replied:
Totally up for learning a new way to do things, but I hope you don't mind my questions-- I'm actively trying to learn and get a deeper understanding of the processes involved. 

     The if changes reverted are for the purpose of making the function more easily understood as you noted, but doesn't that come at the cost of the codes efficiency? Or does nitpicking not do much to help Byond? I've heard it was a finicky engine.

     The decision of no more verbs, is this because verbs have a detrimental impact on performance? Or because of a more firm vision of unified design/intuition to lead the future development? On that note, regarding no verbs, are the altclick and other hotkeys acceptable still?

Thank you so much for taking your time to review this. I hope to hear back! Feel free to add me on Discord if you prefer NOXA#0082

Mothblocks replied:
1. If it's less efficient, then it's a matter of microseconds--we optimize for readability outside of hot paths. 

Quoting the often-cut-off Donald Knuth:

> "Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%." 

That 3% is in places like atmos etc, but not something like tape recorders.

2. Verbs are clunky UX. Right-clicking tiles is often slow, you have to *know* the item has a verb, and players no longer expect it. Hotkeys are completely accepted. Contextual screentips exist so people can simply hover over the item and learn its hotkeys.

NX115 replied:
> Quoting the often-cut-off Donald Knuth:

That was really good. Helped me see things in a new way and it has been kind of popping into my head every so often every time I try and learn and take on a new project — thanks for the wisdom, for real! 

Now, when you say contextual screen tips you're referring primarily to the item description and what the item says on the chatbox when a specific action is done, correct? So you're saying I should allude to features either directly or indirectly as to help give intuition to the player?

Mothblocks reviewed (changes requested):
No more verbs, use contextual screentips to show how to interact with it instead.

`snake_case` only for everything.

No semicolons.

Crushtoe commented:
just adding that the ability to wipe tapes non-destructively makes destroying a tape useless

NX115 committed:
Add wipe icon & remove verbs

Removed verbs as requested, and added an icon for the pop-up menu for wiping the tape in the tape recorder.

NX115 committed:
Remove semicolons & revert 'optimization'

Removed semicolons and no longer squished if statements into ||.

NX115 committed:
Fix bugs.

bug fixing.

NX115 committed:
Final tweak

changes a line of text as to ensure that a player knows what they are doing. Also continues on the previous of requested changes, as I missed a verb.

Mothblocks commented:
@NX115 Didn't see this, if it's ready then undraft it please

NX115 commented:
> @NX115 Didn't see this, if it's ready then undraft it please

How do I go about undrafting it? I haven't quite had to do that before, and I don't really see much regarding it on Google.

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -278,6 +281,16 @@
+				wipeproc()
+
+
+/obj/item/taperecorder/proc/wipeproc()
```
Just name it `wipe` or similar.

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -278,6 +281,16 @@
+	mytape.used_capacity = 0
+	mytape.storedinfo = new
+	mytape.timestamp = new
+	to_chat(usr, "<span class='notice'>You wipe this side of the tape entirely.")
```
Pass the user in directly--we don't use `usr` unless we have to due to its inconsistencies.

Mothblocks replied:
Then, use balloon alerts instead of to_chat, which we use for quick actions that need feedback but not archival.

```suggestion
	balloon_alert(user, "wiped tape")
```

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -393,15 +414,38 @@
 		tapeflip()
 	. = ..()
 
+/obj/item/tape/proc/wipeproc()
```
Name just `wipe`

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -393,15 +414,38 @@
+	. = ..()
+	if (!unspooled)
+		if (firstflip)
+			to_chat(usr, "<span class='notice'>You flip the tape so you can record on the clean magnetic strip.</span>")
```
```suggestion
			balloon_alert(user, "flipped tape to the clean strip")
```

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -393,15 +414,38 @@
+			to_chat(usr, "<span class='notice'>You flip the tape so you can record on the clean magnetic strip.</span>")
+			firstflip = FALSE
+		else
+			to_chat(usr, "<span class='notice'>You flip the tape back around.</span>")
```
```suggestion
			balloon_alert(user, "flipped back around")
```

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -393,15 +414,38 @@
+		tapeflip()
+
+	if (unspooled)
+		to_chat(usr, "<span class='notice'>You scrub the magnetic strip clean of its contents.")
```
```suggestion
		balloon_alert(user, "scrubbed tape")
```

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -393,15 +414,38 @@
+	if (unspooled)
+		cut_overlay("ribbonoverlay")
+		unspooled = FALSE
+	return
```
No trailing returns, remove

Mothblocks commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -393,15 +414,38 @@
 		add_overlay("ribbonoverlay")
-	unspooled = TRUE
+		unspooled = TRUE
+	return
```
No trailing returns, remove

NX115 committed:
Update code/game/objects/items/devices/taperecorder.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

NX115 committed:
Update code/game/objects/items/devices/taperecorder.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

NX115 committed:
Refactor code as requested.

Wipeproc was changed to wipe.
Removed trailing returns.
Removed all 'to_chat' and replaced them with baloons.

NX115 committed:
Bugfix, defining 'user' variable

Forgot to define the user variable. Did that here.
